### PR TITLE
Pass the latest not cancelled IHP task on an appeal to the front end

### PR DESCRIPTION
**Pull request by hschallhorn**
Bumps #15405

### Description
Provides information about ihp tasks on an appeal to the front end

### Acceptance Criteria
- [x] Request date and received date for the latest ihp task on appeal are returned to the front end
- [x] No N+1 queries

### Testing Plan
- [x] Request date and received date for the latest ihp task on appeal are returned to the front end
1. Log in as bvacasper1 and go to their queue
1. Select 4 tasks that meet the following criteria and grab their uuids and vacols_ids
   1. Ama appeal in the "assigned" tab
   1. Ama appeal in the "on hold" tab
   1. Legacy appeal in the "assigned" tab
   1. Legacy appeal in the "on hold" tab
1. Create ihp tasks for these appeals
```ruby
uuids = %w"35de5aee-3243-437d-963f-ae6be85d7152 a33c8a57-92e5-4175-9129-6dc9cccd4be4"
vacols_ids = %w"3626186 2249056"

uuids.each do |uuid|
  appeal = Appeal.find_by(uuid: uuid)
  IhpColocatedTask.create(appeal: appeal, parent: appeal.root_task, assigned_by: User.first, assigned_to: Colocated.singleton)
end

vacols_ids.each do |vacols_id|
  appeal = LegacyAppeal.find_by(vacols_id: vacols_id)
  IhpColocatedTask.create(appeal: appeal, assigned_by: User.first, assigned_to: Colocated.singleton)
end
```
4. Reload the user's queue and check out the dev console
1. Confirm `latest_informal_hearing_presentation_task` is serialized for the 3 ama tasks by selecting the response from "/tasks?..." clicking into queue config, selecting the tab the task should appear in, and looking at the task attributes of the returned task
Screen Shot 2020-11-02 at 1 36 32 PM
1. Confirm `latest_informal_hearing_presentation_task` is serialized for the one legacy tasks by selecting the response from "/BVACASPER1" and looking at the task attributes of the returned task
Screen Shot 2020-11-02 at 1 38 28 PM
1. Complete one of the IHP tasks and ensure the receive date is non null on the front end

```ruby
task = IhpColocatedTask.last
task.completed!
task.appeal
```
Screen Shot 2020-11-02 at 1 41 39 PM

- [x] No N+1 queries (Ensure we're pulling all tasks by all appeal ids, not all tasks one at a time with only one appeal id)
```ruby
QueueConfig.new(assignee: User.first).to_hash
# CTRL-F for ihp
[2020-11-02 15:13:42 -0500]   Task Load (1.1ms)  SELECT "tasks".* FROM "tasks" WHERE "tasks"."status" != $1 AND "tasks"."type" IN ($2, $3) AND "tasks"."appeal_type" = $4 AND "tasks"."appeal_id" IN ($5, $6, $7, $8, $9, $10) ORDER BY "tasks"."closed_at" DESC, "tasks"."assigned_at" DESC  [["status", "cancelled"], ["type", "InformalHearingPresentationTask"], ["type", "IhpColocatedTask"], ["appeal_type", "LegacyAppeal"], ["appeal_id", 1], ["appeal_id", 2], ["appeal_id", 3], ["appeal_id", 4], ["appeal_id", 5], ["appeal_id", 46]]
[2020-11-02 15:13:42 -0500]   Task Load (0.8ms)  SELECT "tasks".* FROM "tasks" WHERE "tasks"."status" != $1 AND "tasks"."type" IN ($2, $3) AND "tasks"."appeal_type" = $4 AND "tasks"."appeal_id" IN ($5, $6, $7, $8, $9, $10, $11, $12) ORDER BY "tasks"."closed_at" DESC, "tasks"."assigned_at" DESC  [["status", "cancelled"], ["type", "InformalHearingPresentationTask"], ["type", "IhpColocatedTask"], ["appeal_type", "Appeal"], ["appeal_id", 30], ["appeal_id", 68], ["appeal_id", 855], ["appeal_id", 856], ["appeal_id", 857], ["appeal_id", 858], ["appeal_id", 859], ["appeal_id", 860]]
```
Example of what an N+1 query would have looked like (remove `:latest_informal_hearing_presentation_task` from task_includes in queue tab)
```ruby
[2020-11-02 15:19:14 -0500]   Task Load (8.5ms)  SELECT  "tasks".* FROM "tasks" WHERE "tasks"."appeal_id" = $1 AND "tasks"."status" != $2 AND "tasks"."type" IN ($3, $4) AND "tasks"."appeal_type" = $5 ORDER BY "tasks"."closed_at" DESC, "tasks"."assigned_at" DESC LIMIT $6  [["appeal_id", 417], ["status", "cancelled"], ["type", "InformalHearingPresentationTask"], ["type", "IhpColocatedTask"], ["appeal_type", "Appeal"], ["LIMIT", 1]]
[2020-11-02 15:19:14 -0500]   Task Load (1.7ms)  SELECT  "tasks".* FROM "tasks" WHERE "tasks"."appeal_id" = $1 AND "tasks"."status" != $2 AND "tasks"."type" IN ($3, $4) AND "tasks"."appeal_type" = $5 ORDER BY "tasks"."closed_at" DESC, "tasks"."assigned_at" DESC LIMIT $6  [["appeal_id", 420], ["status", "cancelled"], ["type", "InformalHearingPresentationTask"], ["type", "IhpColocatedTask"], ["appeal_type", "Appeal"], ["LIMIT", 1]]
[2020-11-02 15:19:14 -0500]   Task Load (3.1ms)  SELECT  "tasks".* FROM "tasks" WHERE "tasks"."appeal_id" = $1 AND "tasks"."status" != $2 AND "tasks"."type" IN ($3, $4) AND "tasks"."appeal_type" = $5 ORDER BY "tasks"."closed_at" DESC, "tasks"."assigned_at" DESC LIMIT $6  [["appeal_id", 373], ["status", "cancelled"], ["type", "InformalHearingPresentationTask"], ["type", "IhpColocatedTask"], ["appeal_type", "Appeal"], ["LIMIT", 1]]
[2020-11-02 15:19:14 -0500]   Task Load (1.0ms)  SELECT  "tasks".* FROM "tasks" WHERE "tasks"."appeal_id" = $1 AND "tasks"."status" != $2 AND "tasks"."type" IN ($3, $4) AND "tasks"."appeal_type" = $5 ORDER BY "tasks"."closed_at" DESC, "tasks"."assigned_at" DESC LIMIT $6  [["appeal_id", 414], ["status", "cancelled"], ["type", "InformalHearingPresentationTask"], ["type", "IhpColocatedTask"], ["appeal_type", "Appeal"], ["LIMIT", 1]]
[2020-11-02 15:19:14 -0500]   Task Load (0.9ms)  SELECT  "tasks".* FROM "tasks" WHERE "tasks"."appeal_id" = $1 AND "tasks"."status" != $2 AND "tasks"."type" IN ($3, $4) AND "tasks"."appeal_type" = $5 ORDER BY "tasks"."closed_at" DESC, "tasks"."assigned_at" DESC LIMIT $6  [["appeal_id", 411], ["status", "cancelled"], ["type", "InformalHearingPresentationTask"], ["type", "IhpColocatedTask"], ["appeal_type", "Appeal"], ["LIMIT", 1]]
[2020-11-02 15:19:14 -0500]   Task Load (1.6ms)  SELECT  "tasks".* FROM "tasks" WHERE "tasks"."appeal_id" = $1 AND "tasks"."status" != $2 AND "tasks"."type" IN ($3, $4) AND "tasks"."appeal_type" = $5 ORDER BY "tasks"."closed_at" DESC, "tasks"."assigned_at" DESC LIMIT $6  [["appeal_id", 364], ["status", "cancelled"], ["type", "InformalHearingPresentationTask"], ["type", "IhpColocatedTask"], ["appeal_type", "Appeal"], ["LIMIT", 1]]
[2020-11-02 15:19:14 -0500]   Task Load (0.7ms)  SELECT  "tasks".* FROM "tasks" WHERE "tasks"."appeal_id" = $1 AND "tasks"."status" != $2 AND "tasks"."type" IN ($3, $4) AND "tasks"."appeal_type" = $5 ORDER BY "tasks"."closed_at" DESC, "tasks"."assigned_at" DESC LIMIT $6  [["appeal_id", 367], ["status", "cancelled"], ["type", "InformalHearingPresentationTask"], ["type", "IhpColocatedTask"], ["appeal_type", "Appeal"], ["LIMIT", 1]]
[2020-11-02 15:19:14 -0500]   Task Load (0.9ms)  SELECT  "tasks".* FROM "tasks" WHERE "tasks"."appeal_id" = $1 AND "tasks"."status" != $2 AND "tasks"."type" IN ($3, $4) AND "tasks"."appeal_type" = $5 ORDER BY "tasks"."closed_at" DESC, "tasks"."assigned_at" DESC LIMIT $6  [["appeal_id", 370], ["status", "cancelled"], ["type", "InformalHearingPresentationTask"], ["type", "IhpColocatedTask"], ["appeal_type", "Appeal"], ["LIMIT", 1]]
[2020-11-02 15:19:14 -0500]   Task Load (1.9ms)  SELECT  "tasks".* FROM "tasks" WHERE "tasks"."appeal_id" = $1 AND "tasks"."status" != $2 AND "tasks"."type" IN ($3, $4) AND "tasks"."appeal_type" = $5 ORDER BY "tasks"."closed_at" DESC, "tasks"."assigned_at" DESC LIMIT $6  [["appeal_id", 376], ["status", "cancelled"], ["type", "InformalHearingPresentationTask"], ["type", "IhpColocatedTask"], ["appeal_type", "Appeal"], ["LIMIT", 1]]
[2020-11-02 15:19:14 -0500]   Task Load (0.8ms)  SELECT  "tasks".* FROM "tasks" WHERE "tasks"."appeal_id" = $1 AND "tasks"."status" != $2 AND "tasks"."type" IN ($3, $4) AND "tasks"."appeal_type" = $5 ORDER BY "tasks"."closed_at" DESC, "tasks"."assigned_at" DESC LIMIT $6  [["appeal_id", 379], ["status", "cancelled"], ["type", "InformalHearingPresentationTask"], ["type", "IhpColocatedTask"], ["appeal_type", "Appeal"], ["LIMIT", 1]]
[2020-11-02 15:19:14 -0500]   Task Load (1.0ms)  SELECT  "tasks".* FROM "tasks" WHERE "tasks"."appeal_id" = $1 AND "tasks"."status" != $2 AND "tasks"."type" IN ($3, $4) AND "tasks"."appeal_type" = $5 ORDER BY "tasks"."closed_at" DESC, "tasks"."assigned_at" DESC LIMIT $6  [["appeal_id", 382], ["status", "cancelled"], ["type", "InformalHearingPresentationTask"], ["type", "IhpColocatedTask"], ["appeal_type", "Appeal"], ["LIMIT", 1]]
[2020-11-02 15:19:14 -0500]   Task Load (0.6ms)  SELECT  "tasks".* FROM "tasks" WHERE "tasks"."appeal_id" = $1 AND "tasks"."status" != $2 AND "tasks"."type" IN ($3, $4) AND "tasks"."appeal_type" = $5 ORDER BY "tasks"."closed_at" DESC, "tasks"."assigned_at" DESC LIMIT $6  [["appeal_id", 385], ["status", "cancelled"], ["type", "InformalHearingPresentationTask"], ["type", "IhpColocatedTask"], ["appeal_type", "Appeal"], ["LIMIT", 1]]
[2020-11-02 15:19:14 -0500]   Task Load (1.2ms)  SELECT  "tasks".* FROM "tasks" WHERE "tasks"."appeal_id" = $1 AND "tasks"."status" != $2 AND "tasks"."type" IN ($3, $4) AND "tasks"."appeal_type" = $5 ORDER BY "tasks"."closed_at" DESC, "tasks"."assigned_at" DESC LIMIT $6  [["appeal_id", 388], ["status", "cancelled"], ["type", "InformalHearingPresentationTask"], ["type", "IhpColocatedTask"], ["appeal_type", "Appeal"], ["LIMIT", 1]]
[2020-11-02 15:19:14 -0500]   Task Load (1.2ms)  SELECT  "tasks".* FROM "tasks" WHERE "tasks"."appeal_id" = $1 AND "tasks"."status" != $2 AND "tasks"."type" IN ($3, $4) AND "tasks"."appeal_type" = $5 ORDER BY "tasks"."closed_at" DESC, "tasks"."assigned_at" DESC LIMIT $6  [["appeal_id", 391], ["status", "cancelled"], ["type", "InformalHearingPresentationTask"], ["type", "IhpColocatedTask"], ["appeal_type", "Appeal"], ["LIMIT", 1]]
[2020-11-02 15:19:14 -0500]   Task Load (0.7ms)  SELECT  "tasks".* FROM "tasks" WHERE "tasks"."appeal_id" = $1 AND "tasks"."status" != $2 AND "tasks"."type" IN ($3, $4) AND "tasks"."appeal_type" = $5 ORDER BY "tasks"."closed_at" DESC, "tasks"."assigned_at" DESC LIMIT $6  [["appeal_id", 423], ["status", "cancelled"], ["type", "InformalHearingPresentationTask"], ["type", "IhpColocatedTask"], ["appeal_type", "Appeal"], ["LIMIT", 1]]
```

### CODE REVIEW

File: app/models/queue_tab.rb [hunk 1]
```diff
@@ -122,7 +122,7 @@ def on_hold_task_children_and_timed_hold_parents
 
   def task_includes
     [
-      { appeal: [:available_hearing_locations, :claimants, :work_mode] },
+      { appeal: [:available_hearing_locations, :claimants, :work_mode, :latest_informal_hearing_presentation_task] },
```
Comment (hschallhorn) [inline app/models/queue_tab.rb:125]: Eager load our ihptasks

File: app/models/legacy_appeal.rb [hunk 2]
```diff
@@ -31,6 +31,11 @@ class LegacyAppeal < CaseflowRecord
   has_many :claimants, -> { Claimant.none }
   has_one :cached_vacols_case, class_name: "CachedAppeal", foreign_key: :vacols_id, primary_key: :vacols_id
   has_one :work_mode, as: :appeal
+  has_one :latest_informal_hearing_presentation_task, lambda {
+    not_cancelled
+      .order(closed_at: :desc, assigned_at: :desc)
+      .where(type: [InformalHearingPresentationTask.name, IhpColocatedTask.name], appeal_type: LegacyAppeal.name)
+  }, class_name: "Task", foreign_key: :appeal_id
```
Comment (hschallhorn) [inline app/models/legacy_appeal.rb:38]: Allows us to eager load these tasks 🤩

File: app/models/appeal.rb [hunk 3]
```diff
@@ -27,6 +27,11 @@ class Appeal < DecisionReview
   has_one :post_decision_motion
   has_many :record_synced_by_job, as: :record
   has_one :work_mode, as: :appeal
+  has_one :latest_informal_hearing_presentation_task, lambda {
+    not_cancelled
+      .order(closed_at: :desc, assigned_at: :desc)
+      .where(type: [InformalHearingPresentationTask.name, IhpColocatedTask.name], appeal_type: Appeal.name)
+  }, class_name: "Task", foreign_key: :appeal_id
```
Comment (hschallhorn) [inline app/models/appeal.rb:34]: Allows us to eager load these tasks 🤩

File: app/models/serializers/work_queue/legacy_task_serializer.rb [hunk 4]
```diff
@@ -91,4 +91,10 @@ class WorkQueue::LegacyTaskSerializer
   attribute :available_actions do |object, params|
     object.available_actions_unwrapper(params[:user], params[:role])
   end
+
+  attribute :latest_informal_hearing_presentation_task do |object|
+    task = object.appeal.latest_informal_hearing_presentation_task
+
+    task ? { requested_at: task.assigned_at, recieved_at: task.closed_at } : {}
+  end
```
Comment (hschallhorn) [inline app/models/serializers/work_queue/legacy_task_serializer.rb:99]: Ensure `latest_informal_hearing_presentation_task` information is passed to the front end for legacy tasks in judge and attorney queues

File: app/models/serializers/work_queue/task_column_serializer.rb [hunk 5]
```diff
@@ -248,6 +248,16 @@ def self.serialize_attribute?(params, columns)
...
 
+  attribute :latest_informal_hearing_presentation_task do |object, params|
+    columns = [Constants.QUEUE_CONFIG.COLUMNS.TASK_TYPE.name, Constants.QUEUE_CONFIG.COLUMNS.DAYS_WAITING.name]
+
+    if serialize_attribute?(params, columns)
+      task = object.appeal.latest_informal_hearing_presentation_task
+
+      task ? { requested_at: task.assigned_at, recieved_at: task.closed_at } : {}
+    end
+  end
```
Comment (hschallhorn) [inline app/models/serializers/work_queue/task_column_serializer.rb:259]: Ensure `latest_informal_hearing_presentation_task` information is passed to the front end in queue configs

**Review by ajspotts — APPROVED** [review]
Smooth ⛵ on the testing plan lgtm!

**Pull request merged**